File: src/lib/features/categorization/domain/entities/category.dart
/// Loại danh mục — đúng một trong hai, cố định sau khi tạo (FR-004, FR-005).
enum CategoryType {
  income,
  expense;

  String get db => this == CategoryType.income ? 'INCOME' : 'EXPENSE';
  String get label => this == CategoryType.income ? 'Thu' : 'Chi';

  static CategoryType fromDb(String value) =>
      value == 'INCOME' ? CategoryType.income : CategoryType.expense;
}

/// Danh mục dùng chung trong hộ; có thể lồng một cấp dưới một danh mục cha. [T018]
class Category {
  final String id;
  final String householdId;
  final String name;
  final CategoryType type;
  final String? icon;
  final String? parentId;
  final bool isDefault;
  final bool isHidden;
  final String? createdBy;

  /// Mốc sửa đổi cuối — dùng cho cập nhật lạc quan giữa các thành viên
  /// (R13, T050). `null` với dữ liệu cũ/demo chưa có mốc.
  final DateTime? updatedAt;

  const Category({
    required this.id,
    required this.householdId,
    required this.name,
    required this.type,
    this.icon,
    this.parentId,
    this.isDefault = false,
    this.isHidden = false,
    this.createdBy,
    this.updatedAt,
  });

  bool get isSubcategory => parentId != null;

  Category copyWith({String? name, String? icon, bool? isHidden}) => Category(
        id: id,
        householdId: householdId,
        name: name ?? this.name,
        type: type, // bất biến (FR-005)
        icon: icon ?? this.icon,
        parentId: parentId,
        isDefault: isDefault,
        isHidden: isHidden ?? this.isHidden,
        createdBy: createdBy,
        updatedAt: updatedAt,
      );
}

File: src/api/module/category/model/category.go
package model

import (
	"time"

	"household-finance/api/common"

	"github.com/google/uuid"
)

// Category — danh mục Thu/Chi dùng chung trong hộ (data-model 001).
type Category struct {
	common.SQLModel
	HouseholdID uuid.UUID  `json:"household_id" gorm:"type:uuid;not null;index"`
	Name        string     `json:"name" gorm:"not null"`
	Type        string     `json:"type" gorm:"not null"` // INCOME | EXPENSE — bất biến sau tạo (FR-005, biz enforce)
	Icon        *string    `json:"icon"`
	ParentID    *uuid.UUID `json:"parent_id" gorm:"type:uuid"`
	IsDefault   bool       `json:"is_default" gorm:"not null;default:false"`
	IsHidden    bool       `json:"is_hidden" gorm:"not null;default:false"`
	CreatedBy   *uuid.UUID `json:"created_by" gorm:"type:uuid"`
	UpdatedAt   time.Time  `json:"updated_at" gorm:"autoUpdateTime"` // mốc lạc quan (D6)
}

func (Category) TableName() string { return "categories" }

// Tree — danh mục gốc kèm con (response GET /api/categories).
type Tree struct {
	Category
	Children []Category `json:"children"`
}

// BuildTree dựng cây cha → con một cấp, giữ thứ tự đầu vào.
// Con có cha không nằm trong danh sách (vd cha bị ẩn khi lọc) được nâng thành gốc
// để không biến mất khỏi response.
func BuildTree(cats []Category) []Tree {
	present := make(map[uuid.UUID]bool, len(cats))
	for _, c := range cats {
		present[c.ID] = true
	}
	children := map[uuid.UUID][]Category{}
	for _, c := range cats {
		if c.ParentID != nil && present[*c.ParentID] {
			children[*c.ParentID] = append(children[*c.ParentID], c)
		}
	}
	roots := []Tree{}
	for _, c := range cats {
		if c.ParentID == nil || !present[*c.ParentID] {
			kids := children[c.ID]
			if kids == nil {
				kids = []Category{}
			}
			roots = append(roots, Tree{Category: c, Children: kids})
		}
	}
	return roots
}
